lib/src/utils/commands_extension.dart

```diff
@@ -139,168 +176,280 @@ extension CommandsClientExtension on Client {
         threadLastEventId: args.threadLastEventId,
       );
     });
-    addCommand('html', (CommandArgs args) async {
+    addCommand('html', (args, stdout) async {
       final event = <String, dynamic>{
         'msgtype': 'm.text',
         'body': args.msg,
         'format': 'org.matrix.custom.html',
         'formatted_body': args.msg,
       };
-      return await args.room.sendEvent(
+      final room = args.room;
+      if (room == null) {
+        throw RoomCommandException();
+      }
+      return await room.sendEvent(
         event,
         inReplyTo: args.inReplyTo,
         editEventId: args.editEventId,
         txid: args.txid,
       );
     });
-    addCommand('react', (CommandArgs args) async {
+    addCommand('react', (args, stdout) async {
       final inReplyTo = args.inReplyTo;
       if (inReplyTo == null) {
         return null;
       }
-      return await args.room.sendReaction(inReplyTo.eventId, args.msg);
+      final room = args.room;
+      if (room == null) {
+        throw RoomCommandException();
+      }
+      final parts = args.msg.split(' ');
+      final reaction = parts.first.trim();
+      if (reaction.isEmpty) {
+        throw CommandException('You must provide a reaction when using /react');
+      }
+      return await room.sendReaction(inReplyTo.eventId, reaction);
     });
-    addCommand('join', (CommandArgs args) async {
-      await args.room.client.joinRoom(args.msg);
+    addCommand('join', (args, stdout) async {
+      final roomId = await args.client.joinRoom(args.msg);
+      stdout?.write(DefaultCommandOutput(rooms: [roomId]).toString());
       return null;
     });
-    addCommand('leave', (CommandArgs args) async {
-      await args.room.leave();
-      return '';
+    addCommand('leave', (args, stdout) async {
+      final room = args.room;
+      if (room == null) {
+        throw RoomCommandException();
+      }
+      await room.leave();
+      return null;
     });
-    addCommand('op', (CommandArgs args) async {
+    addCommand('op', (args, stdout) async {
+      final room = args.room;
+      if (room == null) {
+        throw RoomCommandException();
+      }
       final parts = args.msg.split(' ');
-      if (parts.isEmpty) {
-        return null;
+      if (parts.isEmpty || !parts.first.isValidMatrixId) {
+        throw CommandException('You must enter a valid mxid when using /op');
       }
       int? pl;
       if (parts.length >= 2) {
         pl = int.tryParse(parts[1]);
+        if (pl == null) {
+          throw CommandException(
+            'Invalid power level ${parts[1]} when using /op',
+          );
+        }
       }
       final mxid = parts.first;
-      return await args.room.setPower(mxid, pl ?? 50);
+      return await room.setPower(mxid, pl ?? 50);
```

Comment:
same here?

Reply:
maybe also the mxid

Reply:
The mxid is already available to the user since it's a user provided argument. I'd consider a MXID rather as standard output in cases where users might e.g. search for an MXID or similar as command output ?